bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/dao/impl/PatientDaoImpl.java

```diff
@@ -100,6 +105,66 @@ public PatientDaoImpl(SessionFactory sessionFactory) {
+                                                              Boolean filterPatientsByLocation, Boolean filterOnAllIdentifiers) {
+
+        validateSearchParams(customAttributeFields, programAttributeFieldName, addressFieldName);
+        PatientResponseMapper patientResponseMapper = new PatientResponseMapper(Context.getVisitService(),new BahmniVisitLocationServiceImpl(Context.getLocationService()));
+
+        List<Patient> patients = getPatientsByNameAndGender(name, gender, length);
+        List<Integer> patientIds = patients.stream().map(patient -> patient.getPatientId()).collect(toList());
+        Map<Object, Object> programAttributes = Context.getService(BahmniProgramWorkflowService.class).getPatientProgramAttributeByAttributeName(patientIds, programAttributeFieldName);
+        Set<Integer> uniquePatientIds = new HashSet<>();
+        List<PatientResponse> patientResponses = patients.stream()
+                .map(patient -> {
+                    if(!uniquePatientIds.contains(patient.getPatientId())) {
+                        PatientResponse patientResponse = patientResponseMapper.map(patient, loginLocationUuid, patientSearchResultFields, addressSearchResultFields,
+                                programAttributes.get(patient.getPatientId()));
+                        uniquePatientIds.add(patient.getPatientId());
+                        return patientResponse;
+                    }else
+                        return null;
+                }).filter(Objects::nonNull)
+                .collect(toList());
+        return patientResponses;
+    }
+
+
+    private List<Patient> getPatientsByNameAndGender(String name, String gender, Integer length) {
+        HibernatePatientDAO patientDAO = new HibernatePatientDAO();
+        patientDAO.setSessionFactory(sessionFactory);
+        List<Patient> patients = new ArrayList<Patient>();
+        String query = LuceneQuery.escapeQuery(name);
+        PersonLuceneQuery personLuceneQuery = new PersonLuceneQuery(sessionFactory);
+        LuceneQuery<PersonName> nameQuery = personLuceneQuery.getPatientNameQueryWithOrParser(query, false);
+        /* person.gender does not work somehow in LuceneQuery, so the dirty way is to filter result with person's gender */
+        // if(gender != null && !gender.isEmpty()){
+        //     nameQuery.include("person.gender", gender);
+        // }
+        List<PersonName> persons = nameQuery.list().stream()
+                                    .filter(
+                                        personName ->
+                                            personName.getPreferred()
+                                            && checkGender(personName.getPerson(), gender)
+                                    ).collect(toList());
+        persons = persons.subList(0, Math.min(length, persons.size()));
+        persons.forEach(person -> patients.add(patientDAO.getPatient(person.getPerson().getPersonId())));
```

Comment:
@angshu @djazayeri we do not want to hit the database just to get the Patients from the Persons. We used the lucence index to get the PersonName's of similar Patients via `personLuceneQuery.getPatientNameQueryWithOrParser`, we can access the Person via `personName.getPerson()` but here are accessing the database to get Patients. Do you have a hint for us on how to get Patients without hitting the DB? Can we simply cast them?

cc @eyalgo (who is picking up on the work done by Jian with Martina)

Reply:
answering this late, I haven't ever used OpenMRS's lucene searching, so I don't know offhand how this works.